### PR TITLE
Warn log deprecation warnings received from server

### PR DESCRIPTION
**Pull request by javanna**
The warnings received via response headers get printed out in a single line:

```
WARNING: request [DELETE http://localhost:9200/index/type/_api] returned 3 warnings:[this is warning number 0],[this is warning number 1],[this is warning number 2]
```

### CODE REVIEW

**Comment by cbuescher** [comment]
I was wondering about the reason why all warnings are printed on a single line. I don't know how people are parsing their logs, but I though having multi-lines would be easier to parse, e.g.

```
WARNING: Received warning: 1) indices query is deprecated. Instead search on the '_index' field
WARNING: Received warning: 2) [template] query is deprecated, use search template api instead
```

Otherwise people would need to split the line at the `x)` indicators somehow to get to the individual warnings? Also, in logs where multiple responses might be interleaved, maybe its possible to log something to tie these together? I don't know how difficult it is to build some id for a response that could be logged to keep those warnings together. 

Just my 2 cents, if one line is prefered this looks good to me.

**Comment by javanna** [comment]
good question @cbuescher I tried to keep it all in a single log line because otherwise it becomes hard to understand which request returned which warnings. I could have a line per warning within the same log line, but I'm afraid that is not parser friendly.

> Otherwise people would need to split the line at the x) indicators somehow to get to the individual warnings?

I don't expect to have to log many warnings for each requests, and I think it makes sense to log them per request, I am not sure people will actually split the line to extract each single warning.

I have added now some more info about the request so one can try and trace the warnings back to the request that originated them: 

```
request [PUT http://localhost:9200/index/type/_api] returned warnings: 1) this is warning number 1 2) this is warning number 2
```

**Comment by javanna** [comment]
retest this please

**Comment by javanna** [comment]
@tlrx you mind having a look too please?

File: client/rest/src/main/java/org/elasticsearch/client/RequestLogger.java [hunk 1]
```diff
@@ -97,6 +103,14 @@ static void logFailedRequest(Log logger, HttpUriRequest request, HttpHost host,
         }
     }
 
+    static String buildWarningLine(HttpUriRequest request, HttpHost host, Header[] warnings) {
```
Comment (tlrx) [inline client/rest/src/main/java/org/elasticsearch/client/RequestLogger.java]: buildWarningMessage instead?

File: client/rest/src/main/java/org/elasticsearch/client/RequestLogger.java [hunk 2]
```diff
@@ -97,6 +103,14 @@ static void logFailedRequest(Log logger, HttpUriRequest request, HttpHost host,
...
+        String line = "request [" + request.getMethod() + " " + host + getUri(request.getRequestLine()) + "] returned warnings:";
```
Comment (tlrx) [inline client/rest/src/main/java/org/elasticsearch/client/RequestLogger.java]: Can you use StringJoiner or StringBuilder?

File: client/rest/src/main/java/org/elasticsearch/client/RequestLogger.java [hunk 3]
```diff
@@ -97,6 +103,14 @@ static void logFailedRequest(Log logger, HttpUriRequest request, HttpHost host,
...
+        for (int i = 0; i < warnings.length; i++) {
+            line += " " + (i+1) + ") " + warnings[i].getValue();
```
Comment (tlrx) [inline client/rest/src/main/java/org/elasticsearch/client/RequestLogger.java]: Do we really need to know the number?

File: client/rest/src/test/java/org/elasticsearch/client/RequestLoggerTests.java [hunk 4]
```diff
@@ -162,4 +132,43 @@ public void testTraceResponse() throws IOException {
...
+        HttpUriRequest request = randomHttpRequest(new URI("/index/type/_api"));
+        String expected = "request [" + request.getMethod() + " " + host + "/index/type/_api] returned warnings:";
+        int numWarnings = randomIntBetween(1, 5);
+        Header[] warnings = new Header[numWarnings];
+        for (int i = 0; i < numWarnings; i++) {
+            String warning = "this is warning number " + i;
+            warnings[i] = new BasicHeader("Warning", warning);
+            expected += " " + (i+1) + ") " + warning;
+        }
+        System.out.println(expected);
```
Comment (tlrx) [inline client/rest/src/test/java/org/elasticsearch/client/RequestLoggerTests.java]: Leftover

**Review by tlrx — CHANGES_REQUESTED** [review]
I left some comments. I'm fine with keeping the warnings on a single line but I'd prefer to have the total number of warnings easily parsable and then each warning enclosed in bracket or something:

```
request [PUT http://localhost:9200/index/type/_api] returned [2] warnings: [this is warning number 1][this is warning number 2]
```

File: client/rest/src/main/java/org/elasticsearch/client/RequestLogger.java (hunk 3 above)
Comment (javanna) [inline client/rest/src/main/java/org/elasticsearch/client/RequestLogger.java]: no it was just for convenience.

File: client/rest/src/main/java/org/elasticsearch/client/RequestLogger.java (hunk 1 above)
Comment (javanna) [inline client/rest/src/main/java/org/elasticsearch/client/RequestLogger.java]: sounds good

**Comment by javanna** [comment]
@tlrx ++ on printing the total number of warnings. on the brackets, not sure they will help cause warnings often contain brackets too :)

**Comment by javanna** [comment]
@tlrx I addressed your comments, thanks

**Review by tlrx — APPROVED** [review]

**Pull request merged**